model/gru.py

```diff
@@ -46,21 +46,36 @@ def init_hidden(self):
             earlier versions use "Variable" to initiate tensor
             variable
         '''
-        self.hidden_layer = tensor(torch.zeros(self.nb_lstm_layers,
+        self.hidden_layer = Variable(torch.zeros(self.nb_lstm_layers,
                     self.batch_size, self.nb_lstm_units))
 
-    def forward(self, batch_data):
+    def forward(self, batch_data, seq_lens):
         '''
             apply the forward function for the model
+            clarifying with example from:
+            towardsdatascience.com/\
+            taming-lstms-variable-sized-mini-batches-and-why-pytorch-is-good-for-your-health-61d35642972e
         '''
-        seq_len = batch_data.size()[1]
-        # check that the gru unit treats the s
-        gru_out, self.hidden = self.model(batch_data, self.hidden)
-        output = self.hidden_to_output(gru_out)
+        self.hidden = self.init_hidden()
+         # pack_padded_sequence so that 
+         # padded items in the sequence won't be shown
+        packed_input = utils.rnn.pack_padded_sequence(
+            batch_data, seq_lens, batch_first=True)
+        # run through the GRU model
+        gru_out, _ = self.model(packed_input, self.hidden)
+        # undo packing operation
+        unpacked_out, _ = utils.rnn.pad_packed_sequence(gru_out, batch_first=True)
+        # [TODO] check to see what this unpacking is doing
+        unpacked_out = unpacked_out.contiguous()
+        unpacked_out = unpacked_out.view(-1, unpacked_out.shape[2])
+        unpacked_out = self.hidden_to_output(unpacked_out)
+        # Add a sigmoid layer
+        sigmoid_out = torch.sigmoid(unpacked_out)
+        output = sigmoid_out.view(self.batch_size, -1, self.output_dim)
         return output
 
     def loss(self, output, label):
```

Comment:
@zhampel I have a placeholder loss function here, but what kind of loss would you recommending using and would be most effective if each label is multi-hot vector
[ 0,1,...1,...0]

and the output has the likelihoods.

Reply:
@catli (sorry it took a few days, finally back from travel) For this I think you want to use a multi-class cross entropy type loss function. It's an extension/generalization of binary cross entropy to multiple classes. You can look [here](https://ml-cheatsheet.readthedocs.io/en/latest/loss_functions.html) for a defn, and maybe [here](https://discuss.pytorch.org/t/multi-class-cross-entropy-loss-function-implementation-in-pytorch/19077) for an example PyTorch implementation. Does that help?